Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (c) 2023, Nicolas Ramz <nicolas.ramz@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <AK/Endian.h>
+#include <AK/FixedArray.h>
+#include <AK/Format.h>
+#include <AK/MemoryStream.h>
+#include <AK/String.h>
+#include <LibGfx/ImageFormats/ILBMLoader.h>
+
+namespace Gfx {
+
+struct FourCC {
+    constexpr FourCC(char const* name)
+    {
+        cc[0] = name[0];
+        cc[1] = name[1];
+        cc[2] = name[2];
+        cc[3] = name[3];
+    }
+
+    bool operator==(FourCC const&) const = default;
+    bool operator!=(FourCC const&) const = default;
+
+    char cc[4];
+};
+
+struct IFF_FILE_HEADER {
+    FourCC form;
+    BigEndian<u32> file_size;
+    FourCC format;
+};
+
+static_assert(AssertSize<IFF_FILE_HEADER, 12>());
+
+struct Chunk {
+    FourCC type;
+    ReadonlyBytes data;
+};
+
+enum CompressionType : u8 {
+    Cmp_None = 0,
+    Cmp_ByteRun = 1
+};
+
+enum MaskType : u8 {
+    Mask_None = 0,
+    Mask_HasMask = 1,
+    Mask_HasTransparentColor = 2,
+    Mask_HasLasso = 3
+};
+
+struct ChunkHeader {
+    FourCC chunk_type;
+    BigEndian<u32> chunk_size;
+};
+
+struct BitmapHeader {
+    BigEndian<u16> width;
+    BigEndian<u16> height;
+    BigEndian<i16> x;
+    BigEndian<i16> y;
+    u8 planes;
+    MaskType mask;
+    CompressionType compression;
+    u8 pad;
+    BigEndian<u16> transparent_color;
+    u8 x_aspect;
+    u8 y_aspect;
+    BigEndian<u16> page_width;
+    BigEndian<u16> page_height;
+};
+
+static_assert(sizeof(BitmapHeader) == 20);
+
+struct ILBMLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        BitmapHeaderDecoded,
+        FirstChunkDecoded,
+        ChunksDecoded,
+        FrameDecoded,
+        BitmapDecoded
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    // points to current chunk
+    ReadonlyBytes chunks_cursor;
+
+    // max number of bytes per plane row
+    u16 pitch;
+
+    // Current file color_table in BGRA8888 format
+    Vector<u32> color_table;
+
+    ByteBuffer pixel_data;
```

Comment:
You're basically storing the image twice, once in `pixel_data` and once in the bitmap. I'm pretty sure it can be avoided